src/cascadia/TerminalControl/TSFInputControl.cpp

```diff
@@ -0,0 +1,380 @@
+    {
+        auto request = args.Request();
+
+        // Get window in screen coordinates, this is the entire window including tabs
+        const auto windowBounds = CoreWindow::GetForCurrentThread().Bounds();
+
+        // Get the cursor position in text buffer position
+        auto cursorArgs = winrt::make_self<CursorPositionEventArgs>();
+        _currentCursorPositionHandlers(*this, *cursorArgs);
+        const COORD cursorPos = { gsl::narrow_cast<SHORT>(cursorArgs->CurrentPosition().X), gsl::narrow_cast<SHORT>(cursorArgs->CurrentPosition().Y) };
+
+        // Get Font Info as we use this is the pixel size for characters in the display
+        auto fontArgs = winrt::make_self<FontInfoEventArgs>();
+        _currentFontInfoHandlers(*this, *fontArgs);
+
+        const float fontWidth = fontArgs->FontSize().Width;
+        const float fontHeight = fontArgs->FontSize().Height;
+
+        // Convert text buffer cursor position to client coordinate position within the window
+        COORD clientCursorPos;
+        COORD screenCursorPos;
+        THROW_IF_FAILED(ShortMult(cursorPos.X, gsl::narrow<SHORT>(fontWidth), &clientCursorPos.X));
+        THROW_IF_FAILED(ShortMult(cursorPos.Y, gsl::narrow<SHORT>(fontHeight), &clientCursorPos.Y));
+
+        // Convert from client coordinate to screen coordinate by adding window position
+        THROW_IF_FAILED(ShortAdd(clientCursorPos.X, gsl::narrow_cast<SHORT>(windowBounds.X), &screenCursorPos.X));
+        THROW_IF_FAILED(ShortAdd(clientCursorPos.Y, gsl::narrow_cast<SHORT>(windowBounds.Y), &screenCursorPos.Y));
+
+        // get any offset (margin + tabs, etc..) of the control within the window
+        const auto offsetPoint = this->TransformToVisual(nullptr).TransformPoint(winrt::Windows::Foundation::Point(0, 0));
+
+        // add the margin offsets if any
+        const auto currentMargin = this->Margin();
+        THROW_IF_FAILED(ShortAdd(screenCursorPos.X, gsl::narrow_cast<SHORT>(offsetPoint.X), &screenCursorPos.X));
+        THROW_IF_FAILED(ShortAdd(screenCursorPos.Y, gsl::narrow_cast<SHORT>(offsetPoint.Y), &screenCursorPos.Y));
+
+        // Get scale factor for view
+        const double scaleFactor = DisplayInformation::GetForCurrentView().RawPixelsPerViewPixel();
+
+        // Set the selection layout bounds
+        Rect selectionRect = Rect(screenCursorPos.X, screenCursorPos.Y, 0, fontHeight);
+        request.LayoutBounds().TextBounds(ScaleRect(selectionRect, scaleFactor));
+
+        // Set the control bounds of the whole control
+        Rect controlRect = Rect(screenCursorPos.X, screenCursorPos.Y, 0, fontHeight);
+        request.LayoutBounds().ControlBounds(ScaleRect(controlRect, scaleFactor));
+
+        // position textblock to cursor position
+        _canvas.SetLeft(_textBlock, clientCursorPos.X);
+        _canvas.SetTop(_textBlock, static_cast<double>(clientCursorPos.Y) + 2); // TODO GitHub #3642: Need a better way to align text in center

```

Comment:
When the cursor is near the right side or bottom of the window, does the canvas appear hanging off the side of the window, or is it clipped by the bounds of the window?

If it's clipped, should we file a follow up to try and position the canvas heuristically based on the cursor position relative to the bounds of the window, to keep it in bounds?

Reply:
Opened issue #3657 

_textBlock is clipped by window and text input continues.  This is the same behavior as current command-line, but I agree Terminal should do better.

Reply:
Okay, that sounds fine to me. Thanks for following up!